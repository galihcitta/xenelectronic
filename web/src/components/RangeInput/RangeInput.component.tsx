import React, { ChangeEvent } from 'react';

import { InputWrapper, Label, Info, Input } from './RangeInput.styles';

export interface RangeInputProps {
  name: string;
  value: number;
  onChange: (e: ChangeEvent<HTMLInputElement>) => void;
  min: number;
  max: number;
  id: string;
  label: string;
  info?: string;
}

export const RangeInput: React.FC<RangeInputProps> = ({
  name,
  value,
  onChange,
  min,
  max,
  id,
  label,
  info
}) => {
  return (
    <InputWrapper>
      <Label htmlFor={id}>{label}</Label>
      <Input
        type="range"
        name={name}
        value={value}
        onChange={onChange}
        min={min}
        max={max}
        id={id}
      />
      {info && <Info>{info}</Info>}
    </InputWrapper>
  );
};
