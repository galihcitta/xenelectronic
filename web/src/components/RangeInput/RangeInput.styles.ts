import styled from 'styled-components';

import { fontWeight, device, color } from '../../utils/styles';

export const InputWrapper = styled.div`
  width: 100%;
  margin-bottom: 1.3em;

  @media ${device.laptop} {
    margin-bottom: 0;
  }

  input[type='range'] {
    -webkit-appearance: none;
    background: none;
  }

  input[type='range']::-webkit-slider-runnable-track {
    height: 5px;
    background: #ddd;
    border: none;
    border-radius: 3px;
  }

  input[type='range']::-ms-track {
    height: 5px;
    background: #ddd;
    border: none;
    border-radius: 3px;
  }

  input[type='range']::-moz-range-track {
    height: 5px;
    background: #ddd;
    border: none;
    border-radius: 3px;
  }

  input[type='range']::-webkit-slider-thumb {
    -webkit-appearance: none;
    border: none;
    height: 16px;
    width: 16px;
    border-radius: 50%;
    background: ${color.black};
    margin-top: -5px;
    position: relative;
  }

  input[type='range']::-ms-thumb {
    -webkit-appearance: none;
    border: none;
    height: 16px;
    width: 16px;
    border-radius: 50%;
    background: ${color.black};
    margin-top: -5px;
    position: relative;
  }

  input[type='range']::-moz-range-thumb {
    -webkit-appearance: none;
    border: none;
    height: 16px;
    width: 16px;
    border-radius: 50%;
    background: ${color.black};
    margin-top: -5px;
    position: relative;
  }

  input[type='range']:focus {
    outline-color: ${color.gray};
    &::-webkit-slider-thumb:after {
      position: absolute;
      top: -35px;
      left: 50%;
      transform: translateX(-50%);
      background: #eee;
      border-radius: 5px;
      color: ${color.black};
      padding: 5px 10px;
      border: 2px solid ${color.black};
    }
    &::-ms-thumb:after {
      position: absolute;
      top: -35px;
      left: 50%;
      transform: translateX(-50%);
      background: #eee;
      border-radius: 5px;
      color: ${color.black};
      padding: 5px 10px;
      border: 2px solid ${color.black};
    }
    &::-moz-range-thumb:after {
      position: absolute;
      top: -35px;
      left: 50%;
      transform: translateX(-50%);
      background: #eee;
      border-radius: 5px;
      color: ${color.black};
      padding: 5px 10px;
      border: 2px solid ${color.black};
    }
  }
`;

export const Input = styled.input`
  width: 100%;
  margin-top: 1.5em;
`;

export const Label = styled.label`
  display: block;
  font-size: 0.95rem;
  font-weight: ${fontWeight.semiBold};
  text-align: start;
`;

export const Info = styled.span`
  margin-top: 0.2em;
  font-size: 0.8rem;
`;
