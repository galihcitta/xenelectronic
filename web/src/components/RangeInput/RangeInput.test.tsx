import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { RangeInput, RangeInputProps } from './RangeInput.component';

describe('<RangeInput />', () => {
  const mockOnChange = jest.fn();

  const props: RangeInputProps = {
    name: 'price',
    value: 0,
    onChange: mockOnChange,
    min: 0,
    max: 100,
    id: 'price',
    label: 'price'
  };

  const setup = (props: RangeInputProps) => {
    const utils = render(<RangeInput {...props} />);
    const input = utils.container.querySelector(
      'input[name="price"]'
    ) as HTMLInputElement;

    return { ...utils, input };
  };

  it('renders correctly', () => {
    const { asFragment } = setup(props);
    expect(asFragment()).toMatchSnapshot();
  });

  it('calls `onChange` prop', () => {
    const { input } = setup(props);

    fireEvent.change(input, { target: { value: '50' } });
    expect(mockOnChange).toHaveBeenCalledTimes(1);
  });

  it('renders `Info` element', () => {
    const { getByText } = setup({ ...props, info: 'info message' });
    const info = getByText(/info message/i);

    expect(info).toBeInTheDocument();
    expect(info).toHaveTextContent('info message');
  });
});
