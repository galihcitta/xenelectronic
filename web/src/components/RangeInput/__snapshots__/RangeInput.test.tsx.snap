// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RangeInput /> renders correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bdVaJa iBRfwy"
  >
    <label
      class="sc-htpNat huOBHo"
      for="price"
    >
      price
    </label>
    <input
      class="sc-bwzfXH gcNwFy"
      id="price"
      max="100"
      min="0"
      name="price"
      type="range"
      value="0"
    />
  </div>
</DocumentFragment>
`;
